### PR TITLE
PickList and ValueList: share more code between lists 461

### PR DESCRIPTION
**Pull request by pr3tori4n**
**Related Issue:** #461, #400  

## Summary

TL;DR; I've abstracted some shared code for PickList and ValueList.
In working on a couple bugs for the lists, I've realized that they are already starting to diverge in undesirable ways. From the get-go the test suite was more extensive on pick-list than value-list. And there was quite a bit of duplicated code. I left them isolated since it seemed like these might diverge more based on information from @asangma and @kat10140 but so far they are still very similar.

## Changes

* Several methods have been abstracted to a new filed called `share-list-logic.ts` and both components are importing that as a dependency and use those functions for several internal methods.
* Several tests have been abstracted into `shared-list-tests.ts`, mainly the selection/deselection tests as well as the filtering logic.

## Future
I think it might be possible to also share render logic, but that would need to be abstracted into a functionalComponent at minimum. It might also be possible using higher-order component patterns to share more of the class configuration (props, states and other decorators). Going to pause on that rabbit hole for now so I can get these other bug fixes shipped first.

**_NOTE:_** This is being merged into one of my bug fix branches because I wrote the new filter tests for that bug, then decided to do this rather than copy all of those tests over.

### CODE REVIEW

File: src/components/calcite-pick-list/calcite-pick-list.tsx [hunk 1]
```diff
@@ -53,7 +54,7 @@ export class CalcitePickList {
    */
   @Prop({ reflect: true }) mode: "selection" | "configuration" = "selection";
   /**
-   * Multiple works like conventional checkboxes and radio buttons.
+   * Multiple Works similar to standard radio buttons and checkboxes.
```
Comment (pr3tori4n) [inline src/components/calcite-pick-list/calcite-pick-list.tsx]: The wording here was different from the valueList. Didn't see a good reason for these to use different language.

File: src/components/calcite-pick-list/calcite-pick-list.tsx [hunk 2]
```diff
@@ -104,14 +105,16 @@ export class CalcitePickList {
   //
   // --------------------------------------------------------------------------
 
+  connectedCallback() {
+    sharedListMethods.initialize.call(this);
```
Comment (pr3tori4n) [inline src/components/calcite-pick-list/calcite-pick-list.tsx]: Moving these back to connectedCallback was intentional. Once the logic is modified to use props instead of attributes, it shoouldn't need to be delayed until after first-render. #400

File: src/components/calcite-pick-list/shared-list-logic.ts [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+      if (this.multiple && shiftPressed) {
+        this.selectSiblings(item);
+      }
+      selectedValues.set(value, item);
+    } else {
+      selectedValues.delete(value);
+    }
+    this.lastSelectedItem = item;
+    this.calciteListChange.emit(selectedValues);
+    // this.calcitePickListSelectionChange.emit(selectedValues); picklist only
```
Comment (pr3tori4n) [inline src/components/calcite-pick-list/shared-list-logic.ts]: remove

File: src/components/calcite-tip-manager/calcite-tip-manager.spec.ts [hunk 4]
```diff
@@ -1,4 +1,4 @@
-import { CalciteTipManager } from "./calcite-tip-manager";
+// import { CalciteTipManager } from "./calcite-tip-manager";
```
Comment (pr3tori4n) [inline src/components/calcite-tip-manager/calcite-tip-manager.spec.ts]: I turned on squigglies and started fixing some things :)

File: src/components/calcite-value-list/calcite-value-list.e2e.ts [hunk 5]
```diff
@@ -1,49 +1,14 @@
 import { newE2EPage } from "@stencil/core/testing";
-import { CSS, ICON_TYPES } from "./resources";
+import { ICON_TYPES } from "./resources";
```
Comment (pr3tori4n) [inline src/components/calcite-value-list/calcite-value-list.e2e.ts]: I turned on squigglies and started fixing some things :)

File: src/components/calcite-pick-list/calcite-pick-list.tsx [hunk 6]
```diff
@@ -11,7 +11,8 @@ import {
   h
 } from "@stencil/core";
 import { ICON_TYPES, TEXT } from "./resources";
-import { VNode } from "@esri/calcite-components/dist/types/stencil.core";
+import sharedListMethods from "./shared-list-logic";
```
Comment (driskull) [inline src/components/calcite-pick-list/calcite-pick-list.tsx]: nitpick: might be nicer to import the methods directly as needed instead of having to call `sharedListMethods.method` every time they are used.

`import {method} from "..."`

File: src/components/calcite-pick-list/shared-list-tests.ts [hunk 7]
```diff
@@ -0,0 +1,229 @@
+import { E2EElement, E2EPage, newE2EPage } from "@stencil/core/testing";
+
+export const tests = {
+  selectionAndDeselection(listType) {
```
Comment (jcfranco) [inline src/components/calcite-pick-list/shared-list-tests.ts]: We should try and provide the type for anything we can.

File: src/components/calcite-pick-list/shared-list-tests.ts [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+        filterInput.dispatchEvent(new Event("input"));
+      });
+      await item1.waitForNotVisible();
+
+      item1Visible = await item1.isVisible();
+      item2Visible = await item2.isVisible();
+      expect(item1Visible).toBe(false);
+      expect(item2Visible).toBe(true);
+    });
+    it("should match text in the text-description prop", async () => {
```
Comment (jcfranco) [inline src/components/calcite-pick-list/shared-list-tests.ts]: Nitpick: Adding newlines between tests/suites would improve readability.

File: src/components/calcite-tip-manager/calcite-tip-manager.spec.ts (hunk 4 above)
Comment (jcfranco) [reply src/components/calcite-tip-manager/calcite-tip-manager.spec.ts]: This should be fixed or removed. #nodeadcomments

**Comment by pr3tori4n** [comment]
@driskull I could move to a pure function approach instead of binding the scope, but that would require a good deal more effort refactoring. I was worried I'd have to do it that way, but was able to get it working with this approach, which allowed me to keep the code mostly the same. Is there a solid benefit you feel the functional approach would provide that would be worth that extra effort?

File: src/components/calcite-value-list/calcite-value-list.tsx [hunk 9]
```diff
@@ -132,23 +129,11 @@ export class CalciteValueList {
...
-      if (this.multiple && shiftPressed) {
-        this.selectSiblings(item);
-      }
-      selectedValues.set(value, item);
-    } else {
-      selectedValues.delete(value);
-    }
-    this.lastSelectedItem = item;
-    this.calciteListChange.emit(selectedValues);
+    sharedListMethods.calciteListItemChangeHandler.call(this, event);
```
Comment (jcfranco) [inline src/components/calcite-value-list/calcite-value-list.tsx]: Suggestion: Adding a `thisArg` to the shared methods may help clean up consuming code and also makes it clear that the method needs a different context.

File: .gitignore [hunk 10]
```diff
@@ -24,3 +24,7 @@ Thumbs.db
 UserInterfaceState.xcuserstate
 .env
 src/**/components.d.ts
+
+# backup files
+*.bak
```
Comment (jcfranco) [inline .gitignore]: Where are these files coming from? If not from an IDE, I'd suggest using git stash or local branches.

**Comment by driskull** [comment]
> Is there a solid benefit you feel the functional approach would provide that would be worth that extra effort?

The issue I see is that `this` is typed as `any` so it could be error prone to have it like this. So in the shared code file, anytime `this` is used the typing is lost.

image

File: src/components/calcite-pick-list/shared-list-logic.ts [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+    this.setUpFilter();
+  }),
+  // LifeCycle functions
+  initialize() {
+    // connectedCallback
+    this.setUpItems();
+    this.setUpFilter();
+  },
+  initializeObserver() {
+    // componentDidLoad
```
[the remark this answers is not in the document]
Comment (pr3tori4n) [reply src/components/calcite-pick-list/shared-list-logic.ts]: The comments were mainly for my own benefit during the refactor. I can remove.
My hesitation to match them is that we may need to move them to a different lifecycle in the future, then we'd either have inconsistencies or have to rename these. I figured better to describe the function with its name. But if you feel strongly, I can modify.

[the remark this answers is not in the document]
Comment (jcfranco) [reply src/components/calcite-pick-list/shared-list-logic.ts]: I'm fine with dropping comments if it's not specifically tied to a lifecycle hook. The method names and logic are expressive enough.

File: src/components/calcite-pick-list/shared-list-logic.ts [hunk 12]
```diff
@@ -0,0 +1,110 @@
...
+      obj.label = item.textLabel || item.textHeading;
+      obj.description = item.textDescription;
+      obj.metadata = item.metadata;
+      obj.value = item.value;
+      result.push(obj);
+    });
+    return result;
+  }
+};
+export default sharedListMethods;
```
[the remark this answers is not in the document]
Comment (pr3tori4n) [reply src/components/calcite-pick-list/shared-list-logic.ts]: Read. I'm on board with treating named exports as the best practice and preferred going forward. I hesitate to say "never" use default exports. There may be a time and place for them.
👍

File: .gitignore (hunk 10 above)
Comment (pr3tori4n) [reply .gitignore]: > I'd suggest using git stash or local branches.
👍 
 I don't feel strongly about keeping it it gitignore -  but its a small benefit to keep these file types ignored. Are you also asking to remove this?

**Comment by pr3tori4n** [comment]
@driskull gotcha. Let me see if I can fix the typings in a simple way. If not I'll look at going with a dependency injection model.

File: .gitignore (hunk 10 above)
Comment (jcfranco) [reply .gitignore]: Not right now. I think in the future this file should be revisited.

**Comment by pr3tori4n** [comment]
@driskull I found a solution. Typescript doesn't infer `this` types from calls to `.bind, .call or .apply` automatically, but you can tell a function what it should expect as acceptable values of `this`
https://stackoverflow.com/questions/54710927/typescript-types-and-bind

File: src/components/calcite-pick-list/shared-list-logic.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (driskull) [reply src/components/calcite-pick-list/shared-list-logic.ts]: Nitpick: For this method and the `cleanUpObserver` method, can you just pass the observer? It would probably be nicer to make these pure functions when they can be pure functions.

File: src/components/calcite-pick-list/shared-list-logic.ts [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+  initializeObserver(this: CalcitePickList | CalciteValueList) {
+    // componentDidLoad
+    this.observer.observe(this.el, { childList: true, subtree: true });
+  },
+  cleanUpObserver(this: CalcitePickList | CalciteValueList) {
+    // componentDidUnload
+    this.observer.disconnect();
+  },
+  // Listeners
+  calciteListItemChangeHandler(this: CalcitePickList | CalciteValueList, event): void {
```
Comment (driskull) [inline src/components/calcite-pick-list/shared-list-logic.ts]: fixme: Does event need to be typed?

File: src/components/calcite-pick-list/shared-list-logic.ts [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+      selectedValues.set(value, item);
+    } else {
+      selectedValues.delete(value);
+    }
+    this.lastSelectedItem = item;
+    this.calciteListChange.emit(selectedValues);
+    // this.calcitePickListSelectionChange.emit(selectedValues); picklist only
+  },
+  // Private Methods
+  setUpItems(this: CalcitePickList | CalciteValueList, tagname): void {
```
Comment (driskull) [inline src/components/calcite-pick-list/shared-list-logic.ts]: fixme: Can you type `tagName`?

File: src/components/calcite-pick-list/shared-list-logic.ts [hunk 15]
```diff
@@ -0,0 +1,110 @@
+export const sharedListMethods = {
+  getObserver: new MutationObserver(() => {
```
[the remark this answers is not in the document]
Comment (pr3tori4n) [reply src/components/calcite-pick-list/shared-list-logic.ts]: refactored a bit so it can be used.

File: src/components/calcite-tip-manager/calcite-tip-manager.spec.ts (hunk 4 above)
Comment (pr3tori4n) [reply src/components/calcite-tip-manager/calcite-tip-manager.spec.ts]: The test is valuable but it can't be ran yet because of a Stencil bug. I can't skip the import so have to comment it or remove it. I'd prefer to keep it commented as the test will be easier to turn back on once the spec tests are working. But if you feel strongly, I'll remove.

File: src/components/calcite-pick-list/calcite-pick-list.tsx (hunk 6 above)
Comment (pr3tori4n) [reply src/components/calcite-pick-list/calcite-pick-list.tsx]: Would you be ok if I just added a destructuring call after the import for now?

Comment (driskull) [reply src/components/calcite-pick-list/calcite-pick-list.tsx]: 👍 Its just a nitpick.

Comment (pr3tori4n) [reply src/components/calcite-pick-list/calcite-pick-list.tsx]: added destructuring statement. Avoiding splitting into multiple exports for now since it doesn't help with tree-shaking (they are all used).

File: src/components/calcite-pick-list/shared-list-logic.ts (hunk 13 above)
Comment (pr3tori4n) [reply src/components/calcite-pick-list/shared-list-logic.ts]: done.

File: src/components/calcite-pick-list/shared-list-logic.ts (hunk 14 above)
Comment (pr3tori4n) [reply src/components/calcite-pick-list/shared-list-logic.ts]: yep.

File: src/components/calcite-pick-list/shared-list-logic.ts [hunk 16]
```diff
@@ -0,0 +1,135 @@
+import { CalcitePickList } from "./calcite-pick-list";
+import { CalciteValueList } from "../calcite-value-list/calcite-value-list";
+
+function updateSelectedValuesMap(
+  selectedValuesMap: Map<string, HTMLCalcitePickListItemElement> | Map<string, HTMLCalciteValueListItemElement>,
```
Comment (driskull) [inline src/components/calcite-pick-list/shared-list-logic.ts]: Nitpick: Maybe you could add a couple types in this file for `Map<string, HTMLCalcitePickListItemElement>` and `Map<string, HTMLCalciteValueListItemElement>` as it's repeated a lot and lengthy.

File: src/components/calcite-pick-list/shared-list-logic.ts [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+  item: HTMLCalcitePickListItemElement | HTMLCalciteValueListItemElement
+): void {
+  const selectedValues = (selectedValuesMap as any) as Map<
```
Comment (driskull) [inline src/components/calcite-pick-list/shared-list-logic.ts]: Question: Anyway to avoid the `any` and `as` used here?

Comment (jcfranco) [reply src/components/calcite-pick-list/shared-list-logic.ts]: https://github.com/Esri/calcite-app-components/issues/492

File: src/components/calcite-pick-list/shared-list-logic.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (jcfranco) [reply src/components/calcite-pick-list/shared-list-logic.ts]: > The default export should be renamed to match the filename. 

This has not been addressed.

> There may be a time and place for them.

You can achieve the same without a default export. This has come up before: https://github.com/Esri/calcite-app-components/pull/33#discussion_r293452795

This can be done in a follow-up PR.

File: src/components/calcite-tip-manager/calcite-tip-manager.spec.ts (hunk 4 above)
Comment (jcfranco) [reply src/components/calcite-tip-manager/calcite-tip-manager.spec.ts]: OK, let's keep this for now. Maybe add a comment with what you described? I know it'll stop me from commenting on the same thing in the future. 😄

**Comment by pr3tori4n** [comment]
> This has not been addressed.

I removed it but forgot to push. It should show in the PR now.

> OK, let's keep this for now. Maybe add a comment with what you described? I know it'll stop me from commenting on the same thing in the future. 😄

There is a comment on line 5, I'll move it up to line 1.
